network/p2p/gossip/handler.go

```diff
@@ -17,53 +19,88 @@ import (
 	Namespace          string
 	TargetResponseSize int
 }
 
-func NewHandler[T Gossipable](
-	set Set[T],
+func NewHandler[T any, U GossipableAny[T]](
+	log logging.Logger,
+	client *p2p.Client,
+	set Set[U],
 	config HandlerConfig,
 	metrics prometheus.Registerer,
-) (*Handler[T], error) {
-	h := &Handler[T]{
+) (*Handler[T, U], error) {
+	h := &Handler[T, U]{
 		Handler:            p2p.NoOpHandler{},
+		log:                log,
+		client:             client,
 		set:                set,
 		targetResponseSize: config.TargetResponseSize,
-		sentN: prometheus.NewCounter(prometheus.CounterOpts{
+		pullGossipSentN: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: config.Namespace,
+			Name:      "pull_gossip_sent_n",
+			Help:      "amount of pull gossip sent (n)",
+		}),
+		pullGossipSentBytes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: config.Namespace,
+			Name:      "pull_gossip_sent_bytes",
+			Help:      "amount of pull gossip sent (bytes)",
+		}),
+		pushGossipReceivedN: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: config.Namespace,
+			Name:      "push_gossip_received_n",
+			Help:      "amount of push gossip received (n)",
+		}),
+		pushGossipReceivedBytes: prometheus.NewCounter(prometheus.CounterOpts{
 			Namespace: config.Namespace,
-			Name:      "gossip_sent_n",
-			Help:      "amount of gossip sent (n)",
+			Name:      "push_gossip_received_bytes",
+			Help:      "amount of push gossip received (n)",
 		}),
-		sentBytes: prometheus.NewCounter(prometheus.CounterOpts{
+		pushGossipSentN: prometheus.NewCounter(prometheus.CounterOpts{
 			Namespace: config.Namespace,
-			Name:      "gossip_sent_bytes",
-			Help:      "amount of gossip sent (bytes)",
+			Name:      "push_gossip_sent_n",
+			Help:      "amount of push gossip sent (n)",
+		}),
+		pushGossipSentBytes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: config.Namespace,
+			Name:      "push_gossip_sent_bytes",
+			Help:      "amount of push gossip sent (n)",
 		}),
 	}
 
 	err := utils.Err(
-		metrics.Register(h.sentN),
-		metrics.Register(h.sentBytes),
+		metrics.Register(h.pullGossipSentN),
+		metrics.Register(h.pullGossipSentBytes),
+		metrics.Register(h.pushGossipReceivedN),
+		metrics.Register(h.pushGossipReceivedBytes),
+		metrics.Register(h.pushGossipSentN),
+		metrics.Register(h.pushGossipSentBytes),
 	)
 	return h, err
 }
 
-type Handler[T Gossipable] struct {
+type Handler[T any, U GossipableAny[T]] struct {
 	p2p.Handler
-	set                Set[T]
+	log                logging.Logger
+	set                Set[U]
+	client             *p2p.Client
 	targetResponseSize int
 
-	sentN     prometheus.Counter
-	sentBytes prometheus.Counter
+	pullGossipSentN         prometheus.Counter
+	pullGossipSentBytes     prometheus.Counter
+	pushGossipReceivedN     prometheus.Counter
+	pushGossipReceivedBytes prometheus.Counter
+	pushGossipSentN         prometheus.Counter
+	pushGossipSentBytes     prometheus.Counter
```

Comment:
We could also just use a `CounterVec` with a `type` label of `("push, "pull")`

Reply:
Probably a good idea.